RecoPixelVertexing/PixelTriplets/interface/HitTripletEDProducerT.h

```diff
@@ -0,0 +1,198 @@
+#include "RecoPixelVertexing/PixelTriplets/interface/OrderedHitTriplets.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/IntermediateHitTriplets.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/LayerTriplets.h"
+
+template <typename T_HitTripletGenerator>
+class HitTripletEDProducerT: public edm::stream::EDProducer<> {
+public:
+  HitTripletEDProducerT(const edm::ParameterSet& iConfig);
+  ~HitTripletEDProducerT() = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  virtual void produce(edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+
+private:
+  edm::EDGetTokenT<IntermediateHitDoublets> doubletToken_;
+
+  edm::RunningAverage localRA_;
+
+  T_HitTripletGenerator generator_;
+
+  const bool produceSeedingHitSets_;
+  const bool produceIntermediateHitTriplets_;
+};
+
+template <typename T_HitTripletGenerator>
+HitTripletEDProducerT<T_HitTripletGenerator>::HitTripletEDProducerT(const edm::ParameterSet& iConfig):
+  doubletToken_(consumes<IntermediateHitDoublets>(iConfig.getParameter<edm::InputTag>("doublets"))),
+  generator_(iConfig, consumesCollector()),
+  produceSeedingHitSets_(iConfig.getParameter<bool>("produceSeedingHitSets")),
+  produceIntermediateHitTriplets_(iConfig.getParameter<bool>("produceIntermediateHitTriplets"))
+{
+  if(!produceIntermediateHitTriplets_ && !produceSeedingHitSets_)
+    throw cms::Exception("Configuration") << "HitTripletEDProducerT requires either produceIntermediateHitTriplets or produceSeedingHitSets to be True. If neither are needed, just remove this module from your sequence/path as it doesn't do anything useful";
+  if(produceSeedingHitSets_)
+    produces<RegionsSeedingHitSets>();
+  if(produceIntermediateHitTriplets_)
+    produces<IntermediateHitTriplets>();
+}
+
+template <typename T_HitTripletGenerator>
+void HitTripletEDProducerT<T_HitTripletGenerator>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("doublets", edm::InputTag("hitPairEDProducer"));
+  desc.add<bool>("produceSeedingHitSets", false);
+  desc.add<bool>("produceIntermediateHitTriplets", false);
+
+  T_HitTripletGenerator::fillDescriptions(desc);
+
+  auto label = T_HitTripletGenerator::fillDescriptionsLabel() + std::string("EDProducer");
+  descriptions.add(label, desc);
+}
+
+template <typename T_HitTripletGenerator>
+void HitTripletEDProducerT<T_HitTripletGenerator>::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<IntermediateHitDoublets> hdoublets;
+  iEvent.getByToken(doubletToken_, hdoublets);
+  const auto& regionDoublets = *hdoublets;
+
+  const SeedingLayerSetsHits& seedingLayerHits = regionDoublets.seedingLayerHits();
+  if(seedingLayerHits.numberOfLayersInSet() < 3) {
+    throw cms::Exception("Configuration") << "HitTripletEDProducerT expects SeedingLayerSetsHits::numberOfLayersInSet() to be >= 3, got " << seedingLayerHits.numberOfLayersInSet();
```

Comment:
Not really indeed (at least now). A type-level enforcement would mean a complete overhaul of how the seeding layers are dealt now. Maybe something to think about in the future.

(personally I dislike repeating the check on every event, but currently that's the easiest thing to do by far)